web3.md

```diff
@@ -1086,19 +1094,71 @@ Collecting Coverage Data
+         <execPhase> EPHASE </execPhase>
+         <program> PGM </program>
+      requires notBool {keccak(PGM) | EPHASE} in_keys(OC)
+      [priority(25)]
+
+
+    rule <k> #initVM ... </k>
+         <opcodeLists> OL => OL [ {keccak(PGM) | EPHASE} <- #getOpcodeList(PGM,SCHED) ] </opcodeLists>
          <execPhase> EPHASE </execPhase>
-         <program> CODE </program>
-      requires notBool {keccak(CODE) | EPHASE} in_keys(OC)
+         <schedule> SCHED </schedule>
+         <program> PGM </program>
+      requires notBool {keccak(PGM) | EPHASE} in_keys(OL)
       [priority(25)]
 
+    syntax OpcodeItem ::= "{" Int "|" OpCode "}"
+
+    syntax List ::= #getOpcodeList( ByteArray, Schedule ) [function]
+ // ----------------------------------------------------------------
+    rule #getOpcodeList(PGM , SCHED) => #parseByteCode(0, #sizeByteArray(PGM), PGM, SCHED)
+
+    syntax List ::= #parseByteCode ( Int, Int, ByteArray, Schedule ) [function]
+ // ---------------------------------------------------------------------------
+    rule #parseByteCode(PCOUNT, SIZE, _, _) => .List
+      requires PCOUNT >=Int SIZE
+    rule #parseByteCode(PCOUNT, SIZE, PGM, SCHED) => ListItem({ PCOUNT | #dasmOpCode(PGM [ PCOUNT ], SCHED) } ) #parseByteCode(PCOUNT +Int #widthOp(#dasmOpCode(PGM [ PCOUNT ], SCHED)), SIZE, PGM, SCHED)
+      requires PCOUNT <Int SIZE
+
     rule <k> #execute ... </k>
          <pc> PCOUNT </pc>
          <execPhase> EPHASE </execPhase>
-         <program> CODE </program>
-         <opcodeCoverage> ... { keccak(CODE) | EPHASE } |-> (PCS (.Set => SetItem(PCOUNT))) ... </opcodeCoverage>
+         <program> PGM </program>
+         <opcodeCoverage> ... { keccak(PGM) | EPHASE } |-> (PCS (.Set => SetItem(PCOUNT))) ... </opcodeCoverage>
       requires notBool PCOUNT in PCS
       [priority(25)]
 
+    syntax KItem ::= "#firefly_getCoverageData"
+ // -------------------------------------------
+    rule <k> #firefly_getCoverageData => #sendResponse ("result": #makeCoverageReport(COVERAGE, PGMS)) ... </k>
+         <opcodeCoverage> COVERAGE </opcodeCoverage>
+         <opcodeLists>    PGMS     </opcodeLists>
+
+    syntax JSON ::= #makeCoverageReport ( Map, Map ) [function]
+ // -----------------------------------------------------------
+    rule #makeCoverageReport (COVERAGE, PGMS) => {"coverages": [#serializeCoverage(keys_list(COVERAGE),COVERAGE)], "programs": [#serializePrograms(keys_list(PGMS),PGMS)] }
+
+    syntax JSONList ::= #serializeCoverage ( List, Map ) [function]
+ // ---------------------------------------------------------------
+    rule #serializeCoverage (.List, _ ) => .JSONList
+    rule #serializeCoverage ((ListItem({ CODEHASH | EPHASE } #as KEY) KEYS), KEY |-> X:Set COVERAGE:Map ) => { Int2String(CODEHASH):{ Phase2String(EPHASE): [List2JSONList(Set2List(X))] }}, #serializeCoverage(KEYS, COVERAGE)
+
+    syntax JSONList ::= #serializePrograms ( List, Map ) [function]
```

Comment:
**Note**: instead of having both `#serializeCoverage` and `#serializePrograms` we could keep only the first rule as `#serializeCoverageMap` if `<opcodeLists>` would store `Sets` instead of `Lists`, but I think that the program counters might not be sorted anymore.